test/NuGet.Clients.Tests/NuGet.SolutionRestoreManager.Test/VsSolutionRestoreServiceTests.cs

```diff
@@ -2387,6 +2388,34 @@ public void ToPackageSpec_TwoProjectReferencesToSameProject_DeduplicatesProjectR
         private delegate void TryGetProjectNamesCallback(string projectPath, out ProjectNames projectNames);
         private delegate bool TryGetProjectNamesReturns(string projectPath, out ProjectNames projectNames);
 
+        [Fact]
+        public void ToPackageSpec_PackageDownloadWithNoVersion_ThrowsException()
+        {
+            // Arrange
+            string packageName = "package";
+            ProjectNames projectName = new ProjectNames(@"n:\path\to\current\project.csproj", "project", "project.csproj", "project", Guid.NewGuid().ToString());
+            var emptyReferenceItems = Array.Empty<VsReferenceItem>();
+            var targetFrameworks = new VsTargetFrameworkInfo2[]
+            {
+                new VsTargetFrameworkInfo2("net5.0",
+                    packageReferences: emptyReferenceItems,
+                    projectReferences: emptyReferenceItems,
+                    packageDownloads: new List<IVsReferenceItem>
+                    { new VsReferenceItem(packageName, new VsReferenceProperties(new []
+                    {
+                        new VsReferenceProperty("Version", null)
+                    }))
+                    },
+                    frameworkReferences: emptyReferenceItems,
+                    projectProperties: Array.Empty<IVsProjectProperty>())
+            };
+            string expected = string.Format(CultureInfo.CurrentCulture, Resources.Error_PackageDownload_OnlyExactVersionsAreAllowed, "", packageName);
+
+            // Assert
+            ArgumentException exception = Assert.Throws<ArgumentException>(() => VsSolutionRestoreService.ToPackageSpec(projectName, targetFrameworks, originalTargetFrameworkstr: string.Empty, msbuildProjectExtensionsPath: string.Empty));
```

Comment:
Since this throws an exception, it means that NuGet won't restore any other package in the project. If this project has never had a successful restore, it means that none of the packages this project references will have intellisense until this user-input error is fixed.

An alternative, is to use the "additional errors" code I added years ago. This means that `ToPackageSpec` should complete without throwing an exception (in this case, I'd leave the bad PackageDownload out of the PackageSpec that's returned), but the project adapter cache should have the error message pre-cached. Then, restore will automatically add the message as an error, but it allows restore to run to set up all the other packages in the project.

I don't know how other people in the team feel about this, but I think NuGet has really bad handling of input errors. Nobody wants a compiler that stops at the first error, without reporting all the other errors in the file, and I don't think Nuget should either.

But I know this would be a lot more complicated for someone just starting in the team to implement.

Reply:
> An alternative, is to use the "additional errors" code I added years ago

This will use the AdditionalErrors code you added. 

https://github.com/NuGet/NuGet.Client/blob/fd095eca62b6d63ba5a6a68cfdfba63c23a243e1/src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs#L221-L242

> Since this throws an exception, it means that NuGet won't restore any other package in the project. If this project has never had a successful restore, it means that none of the packages this project references will have intellisense until this user-input error is fixed.

It's not quite the case with PackageDownload, but restore does not really work in a way where it allows for a partial restore. Any change in the top level packages can massively affect the graph.

Reply:
> Any change in the top level packages can massively affect the graph.

Due to how DependencyResolver works, we still extract all the same packages to the global packages folder. So, it's not really doing any work that wouldn't be done if all errors were fixed. But yes, I concede that the transitive packages can be very different.

> See above Re:Partial restore.
> I don't think this is worth it.

Even if we don't want to skip the packages and do a restore, I still wish that we thought about at least detecting all input errors and then fail the nomination (and equivalent in CLI restores), rather than bailing out on the first error. On my (now broken) Surface Book 2, it took MSBuild ~12 seconds to evaluate all the projects in NuGet.sln. If I have 2 input errors, and I have to wait 12+ seconds each time, rather than once and let me fix both errors, I feel frustrated.

Reply:
> I still wish that we thought about at least detecting all input errors and then fail the nomination (and equivalent in CLI restores), rather than bailing out on the first error

I don't mind this, I'm just not sure this is common to warrant the amount work it'd take :) 

Maybe you can create an issue about this? I think it's something worth tracking, and the best way to prove/disprove the sentiment.

Reply:
Since the agreement is not to make the updates in this PR, I have marked the conversations as resolved. Please let me know if you disagree.